### PR TITLE
[JENKINS-4758] Remove old JDK installer from the plugin

### PR DESCRIPTION
**Pull request by kuisathaverat**
[JENKINS-47586](https://issues.jenkins-ci.org/browse/JENKINS-47586)

* Removes the JDKInstaller and code related
* Refactor of method and classes to manage the JDK version to one class JavaVersionChecker
* Extract DefaultJavaProvider from SSHLauncher

### CODE REVIEW

**Comment by kuisathaverat** [comment]
yep, make sense to set a proper order, on the current implementation use first the PATH and fixed paths than configured values, it seems incorrect.

* Agent working directory - WORKING_DIRECTORY/jdk/bin/java
* JAVA_HOME environment variable - JAVA_HOME/bin/java
* JDK tools configured on Jenkins - JDK_TOOLS_LOCATIONS/bin/java
* PATH
* "/usr/bin/java"
* "/usr/java/default/bin/java"
* "/usr/java/latest/bin/java"
* "/usr/local/bin/java"
* "/usr/local/java/bin/java

File: src/main/java/hudson/plugins/sshslaves/DefaultJavaProvider.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+ */
+@Extension
+public class DefaultJavaProvider extends JavaProvider {
+
+    public static final String JAVA_HOME = "JAVA_HOME";
+    public static final String BIN_JAVA = "/bin/java";
+
+    @Override
+    public List<String> getJavas(SlaveComputer computer, TaskListener listener, Connection connection) {
+        List<String> javas = new ArrayList<>(); // this is where we attempt to auto-install
```
[the remark this answers is not in the document]
Comment (kuisathaverat) [reply src/main/java/hudson/plugins/sshslaves/DefaultJavaProvider.java]: it is a leftover to remove.

File: src/main/java/hudson/plugins/sshslaves/JavaVersionChecker.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    public JavaVersionChecker(SlaveComputer computer, TaskListener listener, String jvmOptions, Connection connection) {
+        this.computer = computer;
+        this.listener = listener;
+        this.jvmOptions = jvmOptions;
+        this.connection = connection;
+    }
+
+    /**
+     * return javaPath if specified in the configuration.
+     * Finds local Java, and if none exist, install one.
```
[the remark this answers is not in the document]
Comment (kuisathaverat) [reply src/main/java/hudson/plugins/sshslaves/JavaVersionChecker.java]: ditto

File: src/main/java/hudson/plugins/sshslaves/JavaVersionChecker.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                LOGGER.fine("Trying Java at " + javaCommand);
+                try {
+                    return checkJavaVersion(listener, javaCommand);
+                } catch (IOException e) {
+                    LOGGER.log(FINE, "Failed to check the Java version",e);
+                    // try the next one
+                }
+            }
+        }
+        throw new IOException("Java not found " + computer + ", install a Java 8+ version on the Agent, if you need "
```
[the remark this answers is not in the document]
Comment (kuisathaverat) [reply src/main/java/hudson/plugins/sshslaves/JavaVersionChecker.java]: you are right, it is confusing, let's keep it simple.

`"Java not found on " + computer + ". Install a Java 8 version on the Agent.`

File: src/main/java/hudson/plugins/sshslaves/JavaVersionChecker.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        if(null == result) {
+            listener.getLogger().println(Messages.SSHLauncher_UnknownJavaVersion(javaCommand));
+            listener.getLogger().println(output);
+            throw new IOException(Messages.SSHLauncher_UnknownJavaVersion(javaCommand));
+        } else {
+            return result;
+        }
+    }
+
+    // XXX switch to standard method in 1.479+
```
[the remark this answers is not in the document]
Comment (kuisathaverat) [reply src/main/java/hudson/plugins/sshslaves/JavaVersionChecker.java]: not at all

**Comment by daniel-beck** [comment]
This should probably come with a compatibleSince warning, as it will result in many agents suddenly failing to start.

**Comment by kuisathaverat** [comment]
@daniel-beck 
I do not think so, remoting requires java 8 in 2.73.3 Jenkins Core, and the installer is broken, in any case, yep, it could be nice to add compatibleSince to this new version, I never used it, I suppose that you will see a warning or something when you see the plugin in the plugin list.

**Comment by daniel-beck** [comment]
> the installer is broken, in any case

… unless you configured Oracle credentials in core?

**Pull request merged**